moveit_ros/occupancy_map_monitor/src/occupancy_map_monitor_middleware_handle.cpp

```diff
@@ -0,0 +1,164 @@
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Tyler Weaver */
+
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include <pluginlib/class_loader.hpp>
+#include <rclcpp/rclcpp.hpp>
+
+#include <moveit/occupancy_map_monitor/occupancy_map_monitor_middleware_handle.hpp>
+#include <moveit/occupancy_map_monitor/occupancy_map_updater.h>
+
+namespace occupancy_map_monitor
+{
+namespace
+{
+static const rclcpp::Logger LOGGER = rclcpp::get_logger("moveit.ros.occupancy_map_monitor.middleware_handle");
+}
+
+OccupancyMapMonitorMiddlewareHandle::OccupancyMapMonitorMiddlewareHandle(const rclcpp::Node::SharedPtr& node,
+                                                                         double map_resolution,
+                                                                         const std::string& map_frame)
+  : node_{ node }, map_resolution_{ map_resolution }, map_frame_{ map_frame }, sensor_plugins_{}
+{
+  try
+  {
+    updater_plugin_loader_ = std::make_unique<pluginlib::ClassLoader<OccupancyMapUpdater>>(
+        "moveit_ros_occupancy_map_monitor", "occupancy_map_monitor::OccupancyMapUpdater");
+  }
+  catch (pluginlib::PluginlibException& ex)
+  {
+    RCLCPP_FATAL_STREAM(LOGGER, "Exception while creating octomap updater plugin loader " << ex.what());
+    throw ex;
+  }
+
+  if (map_resolution_ <= std::numeric_limits<double>::epsilon())
+  {
+    if (!node_->get_parameter("octomap_resolution", map_resolution_))
+    {
+      map_resolution_ = 0.1;
+      RCLCPP_WARN(LOGGER, "Resolution not specified for Octomap. Assuming resolution = %g instead", map_resolution_);
+    }
+  }
+
+  if (map_frame.empty())
+  {
+    node_->get_parameter("octomap_frame", map_frame_);
+  }
+
+  std::vector<std::string> sensor_names;
+  if (!node_->get_parameter("sensors", sensor_names))
+  {
+    RCLCPP_ERROR(LOGGER, "No 3D sensor plugin(s) defined for octomap updates");
+  }
+  else if (sensor_names.empty())
+  {
+    RCLCPP_ERROR(LOGGER, "List of sensors is empty!");
+  }
+
+  for (const auto& sensor_name : sensor_names)
+  {
+    std::string sensor_plugin = "";
+    if (!node_->get_parameter(sensor_name + ".sensor_plugin", sensor_plugin))
+    {
+      RCLCPP_ERROR(LOGGER, "No sensor plugin specified for octomap updater %s; ignoring.", sensor_name.c_str());
+    }
+
+    if (sensor_plugin.empty() || sensor_plugin[0] == '~')
+    {
+      RCLCPP_INFO_STREAM(LOGGER, "Skipping octomap updater plugin '" << sensor_plugin << "'");
+      continue;
+    }
+    else
+    {
+      sensor_plugins_.push_back(std::make_pair(sensor_name, sensor_plugin));
```

Comment:
Sure, generally I avoid emplace_back because it is easy to create unintended situations with it.  However in this case I think it does make sense.

https://abseil.io/tips/112